frontend/apps/reader/modules/readersearch.lua

```diff
@@ -51,33 +83,95 @@ function ReaderSearch:onShowFulltextSearchInput()
                     text = forward_text,
                     is_enter_default = true,
                     callback = function()
                         if self.input_dialog:getInputText() == "" then return end
-                        UIManager:close(self.input_dialog)
-                        self:onShowSearchDialog(self.input_dialog:getInputText(), 0)
+                        self.last_search_text = self.input_dialog:getInputText()
+                        self.use_regex = self.check_button_regex.checked
+                        self.case_insensitive = not self.check_button_case.checked
+                        if self.use_regex and self.ui.document:checkRegex(self.input_dialog:getInputText()) ~= 0 then
+                            UIManager:show(InfoMessage:new{ text = _("Invalid regular expression!") })
+                        else
+                            UIManager:close(self.input_dialog)
+                            self:onShowSearchDialog(self.input_dialog:getInputText(), 0, self.use_regex, self.case_insensitive)
+                        end
                     end,
                 },
             },
         },
     }
+
+   -- checkboxes
+    self.check_button_regex = CheckButton:new{
+        text = _("Regular expression"),
+        checked = self.use_regex,
+        parent = self.input_dialog,
+        callback = function()
+            if not self.check_button_regex.checked then
+                self.check_button_regex:check()
+            else
+                self.check_button_regex:unCheck()
+            end
+        end,
+    }
+    self.check_button_case = CheckButton:new{
+        text = _("Case sensitive"),
+        checked = not self.case_insensitive,
+        parent = self.input_dialog,
+        callback = function()
+            if not self.check_button_case.checked then
+                self.check_button_case:check()
+            else
+                self.check_button_case:unCheck()
+            end
+        end,
+    }
+
+    local checkbox_shift = math.floor((self.input_dialog.width - self.input_dialog._input_widget.width) / 2 + 0.5)
+    local check_buttons = HorizontalGroup:new{
+        HorizontalSpan:new{width = checkbox_shift},
+        VerticalGroup:new{
+            align = "left",
+            self.check_button_case,
+            self.check_button_regex,
+        },
+    }
+
+    -- insert check buttons before the regular buttons
+    local nb_elements = #self.input_dialog.dialog_frame[1]
+    table.insert(self.input_dialog.dialog_frame[1], nb_elements-1, check_buttons)
+
     UIManager:show(self.input_dialog)
     self.input_dialog:onShowKeyboard()
 end
 
-function ReaderSearch:onShowSearchDialog(text, direction)
+function ReaderSearch:onShowSearchDialog(text, direction, regex, case_insensitive)
     local neglect_current_location = false
     local current_page
+
+    local function isSlowRegex(text)
+        if text:find("%[") or text:find("%*") or text:find("%?") or text:find("%.") then
```

Comment:
It doesn't seem that [] should be particularly slow?

Reply:
But the regex `[a-z]` takes long on my Tolino and even a second in the emulator.

Reply:
Fair, although that seems a bit contrived. :-P